### ISSUE TITLE
Adapting VQAScore to a new VLM

### ISSUE THREAD

**Issue by alzaia** [linzhiqiu/t2v_metrics#7]
Hey there, I am interested in running VQAScore with another VLM, CogVLM (see [here](https://huggingface.co/THUDM/cogvlm-chat-hf)). I was looking at the guidelines on how to adapt to another VQA model checking the LLaVA template example. I'm wondering how much work should I expect to put in for this? The code of interest is [here](https://huggingface.co/THUDM/cogvlm-chat-hf/blob/main/modeling_cogvlm.py). Any tips/advice would be appreciated!

**Comment by linzhiqiu** [comment]
Hey, I am not familiar with CogVLM architecture, but it seems like their implementation is fairly clean. 

I would suggest you first check the code for CLIP-FlanT5/LLaVA-1.5 and make sure you understand how we compute the token likelihood (e.g., logits) in the forward() function. Then, it should be fairly simple to adapt CogVLM's [forward() function ](https://huggingface.co/THUDM/cogvlm-chat-hf/blob/e29dc3ba206d524bf8efbfc60d80fc4556ab0e3c/modeling_cogvlm.py#L638) for computing VQAScore.

Please reach out if you need any help.

**Comment by alzaia** [comment]
Makes sense, thank you! I will likely try to adapt it to VLMs I currently playing with.

**Comment by alzaia** [comment]
Hello again, I am currently trying to add CogVLM in the VQAScore codebase. However, I am unsure how much code I need to change/adapt to make it compatible. 

- Do I actually need to add the CogVLM src code under its own directory under `vqascore_models`? I was hoping to make use of the automatic `AutoModel` classes here without adding any src code. 
- How much do I actually need to change in the main file (here `cogvlm_model.py`, which I started from `llava_model.py` since they are very similar). For instance, updating the model name and tokenizer/model ids:
```
COGVLM_MODELS = {
    'cogvlm-17b': {
        'tokenizer' : {
            'path': 'lmsys/vicuna-7b-v1.5',
        },
        'model': {
            'path': 'THUDM/cogvlm-chat-hf',
            'conversation': 'chat',
            'image_aspect_ratio': 'pad',
        },
    },
}
```
Do you have any instructions/tips on what to change in the codebase when adding a new model? E.g. do I need to inherit and reimplement the classes such as ModelArguments (following the example in `llava/llava_llama.py` here)? Any help is greatly appreciated, thanks!

**Comment by linzhiqiu** [comment]
Hey,

1 - You don't need to add CogVLM src code under its directory and I believe you can simply use `AutoModel`. The only purpose I did that for LLaVA/InstructBLIP is because I am afraid their src code might change in the future (at least LLaVA-1.6 is different from LLaVA-1.5 already). 
2 - You can do anything you want for `cogvlm_model.py`. There is no need to follow anything as long as you implemented the CogVLM class (a child class of `VQAScoreModel`). 

The only thing you need to change after you complete your model is to update the [t2v_metrics/models/vqascore_models/__init__.py](t2v_metrics/models/vqascore_models/__init__.py). It should be straightforward to add your models to `ALL_VQA_MODELS` and `get_vqascore_model()`. Then you should be good to go!

**Comment by alzaia** [comment]
Hey @linzhiqiu , thanks for the feedback. Sadly I have not much experience with LLMs in practice, but tried to write a skeleton for the `CogVLM` class, based on your examples for `LLaVA` and `CLIP-T5`, keeping in mind that I want to reuse as much as possible from [the CogVLM code from HuggingFace](https://huggingface.co/THUDM/cogvlm-chat-hf/blob/main/modeling_cogvlm.py). 

This is what I have so far (surprisingly runs but there are clearly issues in the way I process the labels/inputs since it gives me a score of 0):
```
from typing import List
import torch
import copy
from PIL import Image

from .vqa_model import VQAScoreModel
from ...constants import HF_CACHE_DIR
from transformers import AutoModelForCausalLM, LlamaTokenizer

CONTEXT_LEN = 2048
SYSTEM_MSG = ""
IGNORE_INDEX = -100
DEFAULT_IMAGE_TOKEN = "<image>"

default_question_template = 'Does this figure show "{}"? Please answer yes or no.'
default_answer_template = "Yes"

def format_question(question, conversation_style='chat'):
    if conversation_style == 'plain':
        question = DEFAULT_IMAGE_TOKEN + question
    elif conversation_style == 'chat':
        #question = SYSTEM_MSG + " USER: " + DEFAULT_IMAGE_TOKEN + "\n" + question + " ASSISTANT: "
        question = question #not sure about this, from experience CogVLM inference does not require a text template unlike LLaVA
    else:
        raise NotImplementedError()
    return question

def format_answer(answer, conversation_style='chat'):
    if conversation_style == 'plain':
        answer = answer + "\n"
    elif conversation_style == 'chat':
        answer = answer + "</s>"
    else:
        raise NotImplementedError()
    return answer

COGVLM_MODELS = {
    'cogvlm-17b': {
        'tokenizer' : {
            'path': 'lmsys/vicuna-7b-v1.5',
        },
        'model': {
            'path': 'THUDM/cogvlm-chat-hf',
            'conversation': 'chat',
            'image_aspect_ratio': 'pad',
        },
    },
}

class CogVLMModel(VQAScoreModel):
    """A wrapper for the CogVLM model"""
    def __init__(self,
                 model_name='cogvlm-17b',
                 device='cuda',
                 cache_dir=HF_CACHE_DIR):
        assert model_name in COGVLM_MODELS
        super().__init__(model_name=model_name,
                         device=device,
                         cache_dir=cache_dir)

    def load_model(self):
        """Load the model, tokenizer, image transform
        """

        self.conversational_style = COGVLM_MODELS[self.model_name]['model']['conversation']
        
        self.context_len = CONTEXT_LEN

        self.tokenizer = LlamaTokenizer.from_pretrained('lmsys/vicuna-7b-v1.5', cache_dir=HF_CACHE_DIR)
        
        self.model = AutoModelForCausalLM.from_pretrained(
            'THUDM/cogvlm-chat-hf',
            torch_dtype=torch.bfloat16,
            low_cpu_mem_usage=True,
            trust_remote_code=True,
            cache_dir=HF_CACHE_DIR,
            ).to('cuda').eval()

    def load_images(self, image_paths: List[str]) -> torch.Tensor:
        images = [Image.open(path).convert("RGB") for path in image_paths]
        return images

    @torch.no_grad()
    @torch.autocast(device_type='cuda', dtype=torch.bfloat16)
    def forward(self,
                images: List[str],
                texts: List[str],
                question_template: str=default_question_template,
                answer_template: str=default_answer_template) -> torch.Tensor:
        """Forward pass of the model to return n scores for n (image, text) pairs (in PyTorch Tensor)
        """
        assert len(images) == len(texts), "Number of images and texts must match"

        questions = [question_template.format(text) for text in texts]
        answers = [answer_template.format(text) for text in texts]
        
        # Formatting for LLaVA-1.5 desired input including system message and image tokens
        questions = [format_question(question, conversation_style=self.conversational_style) for question in questions]
        answers = [format_answer(answer, conversation_style=self.conversational_style) for answer in answers]
        
        images = self.load_images(images)

        prompts = [qs + ans for qs, ans in zip(questions, answers)]
        print("prompts: ", prompts[0])

        inputs = self.model.build_conversation_input_ids(self.tokenizer, query=prompts[0], history=[], images=[images[0]])

        inputs = {
            'input_ids': inputs['input_ids'].unsqueeze(0).to('cuda'),
            'token_type_ids': inputs['token_type_ids'].unsqueeze(0).to('cuda'),
            'attention_mask': inputs['attention_mask'].unsqueeze(0).to('cuda'),
            'images': [[inputs['images'][0].to('cuda').to(torch.bfloat16)]],
        }
        
        input_ids = inputs['input_ids']
        labels = copy.deepcopy(input_ids)

        for label, qs in zip(labels, questions):
            tokenized_len = len(input_ids)
            if qs[-1] == " ":
                tokenized_len -= 1
            label[:tokenized_len] = IGNORE_INDEX
    
        input_ids = torch.nn.utils.rnn.pad_sequence(
            input_ids,
            batch_first=True,
            padding_value=self.tokenizer.pad_token_id)

        labels = torch.nn.utils.rnn.pad_sequence(labels,
                                                batch_first=True,
                                                padding_value=IGNORE_INDEX)
        
        input_ids = input_ids[:, :self.tokenizer.model_max_length]
        labels = labels[:, :self.tokenizer.model_max_length]
        attention_mask = input_ids.ne(self.tokenizer.pad_token_id)

        input_ids, attention_mask, labels = input_ids.to(self.device), attention_mask.to(self.device), labels.to(self.device)

        outputs = self.model.model(**inputs, return_dict=True)

        hidden_states = outputs[0]
        logits = self.model.lm_head(hidden_states)

        # Shift so that tokens < n predict n
        shift_logits = logits[..., :-1, :].contiguous()
        shift_labels = labels[..., 1:].contiguous()

        # Flatten the tokens
        loss_fct = torch.nn.CrossEntropyLoss(reduction='mean')
        shift_labels = shift_labels.to(shift_logits.device)
        lm_prob = torch.zeros(shift_logits.shape[0])
        for k in range(lm_prob.shape[0]):
            lm_prob[k] = (-loss_fct(shift_logits[k], shift_labels[k])).exp()
        return lm_prob
```

Maybe I should make use of [these lines](https://huggingface.co/THUDM/cogvlm-chat-hf/blob/main/modeling_cogvlm.py#L679) since it does pretty much the same thing? Any pointers/tips would be appreciated.

**Comment by linzhiqiu** [comment]
Yes, I think you should refer to the lines you sent me instead of reusing LLaVA-1.5's code -- I am not sure how much they overlap in fact.

**Comment by alzaia** [comment]
Thanks @linzhiqiu, I did manage to integrate CogVLM into it by making use of the original CogVLM source code from HuggingFace! I currently have a forked version, I probably need to clean up the implementation a bit. But I could eventually submit a PR.

**Comment by linzhiqiu** [comment]
Thank you for the PR! 

Have you tried CogVLM on any datasets? For example, you can run eval.py to evaluate your model on all benchmarks we tested in the paper. To get started, I think Winoground is a good dataset (shouldn't take more than a few minutes). If CogVLM is strong enough, you should see performance close to LLaVA-1.5 or CLIP-FlanT5 (a group score of above 28).

I am recently working on implementing VQAScore using GPT-4-Vision, and I will merge your PR once I am done with that. In the meantime, I would appreciate it if you could test CogVLM on Winoground / other benchmarks that you find suitable.

**Comment by alzaia** [comment]
Hey, I will try to run the eval on the benchmark and clean the code more so I can submit a PR (but may not be able to do so in the next 2 weeks due to vacation). 

I compared LLaVA 13B vs CogVLM vs CLIP-FlanT5-XXL on a binary classification task that I'm working on so far, and it seems that CLIP-FlanT5-XXL is still the best one, CogVLM coming second (for a given caption, P(Yes), using AUC as metric).

**Comment by linzhiqiu** [comment]
Sure! Take your time!
